File: java/AlgorithmAndDataStructure/src/test/java/algorithms/NumericTest.java
package algorithms;

import functionObjects.BinaryOp;
import functionObjects.Itoa;
import functionObjects.PlusOne;
import functionObjects.XPlusY;
import junit.framework.TestCase;

import java.util.ArrayList;
import java.util.Arrays;

public class NumericTest extends TestCase {
	private ArrayList<Integer> a;
	private ArrayList<Integer> b;
	private BinaryOp<Integer> xPlusY;

	public void setUp() throws Exception {
		super.setUp();
		a = new ArrayList<Integer>(Arrays.asList(1, 2, 3));
		b = new ArrayList<Integer>(Arrays.asList(4, 5, 6));
		xPlusY = new XPlusY();
	}

	public void testItoa() throws Exception {
		Itoa<Integer> itoa = new PlusOne(4);
		Numeric.itoa(a, itoa);
		assertEquals(b, a);
	}

	public void testAccumulate() throws Exception {
		assertEquals(Integer.valueOf(6), Numeric.accumulate(a, 0, xPlusY));
	}

	public void testInnerProduct() throws Exception {
		ArrayList<Integer> c = new ArrayList<Integer>(Arrays.asList(5, 7, 9));
		ArrayList<Integer> out = new ArrayList<Integer>(3);
		Numeric.innerProduct(a, b, out, xPlusY);
		assertEquals(c, out);
	}

	public void testAdjacentDifference() throws Exception {
		ArrayList<Integer> c = new ArrayList<Integer>(Arrays.asList(1, 3, 5));
		ArrayList<Integer> out = new ArrayList<Integer>(3);
		Numeric.adjacentDifference(a, out, xPlusY);
		assertEquals(c, out);
	}

	public void testPartialSum() throws Exception {
		ArrayList<Integer> c = new ArrayList<Integer>(Arrays.asList(1, 3, 6));
		ArrayList<Integer> out = new ArrayList<Integer>(3);
		Numeric.partialSum(a, out, xPlusY);
		assertEquals(c, out);
	}
}

File: php/NumericTest.php
<?php
namespace Algorithm;

class NumericTest extends \PHPUnit_Framework_TestCase
{
    private $ary;
    public function setUp() {
        $this->ary = [2, 5, 7];
    }

    public function testItoa()
    {
        $g = itoa(4);
        $this->assertEquals(4, $g());
        $this->assertEquals(5, $g());
    }

    public function testAccumulate()
    {
        $this->assertEquals(1, accumulate([], 1, function($a, $b) {
            return $a * $b;
        }));
        $this->assertEquals(70, accumulate($this->ary, 1, function($a, $b) {
            return $a * $b;
        }));
        $this->assertEquals(140, accumulate($this->ary, 2, function($a, $b) {
            return $a * $b;
        }));
    }

    public function testInnerProduction()
    {
        $this->assertEquals(1, innerProduction([], [], 1));
        // require count(ary1) <= count(ary2)
        $this->assertEquals(32, innerProduction($this->ary, [1, 2, 3, 4], -1));
        $this->assertEquals(79, innerProduction($this->ary, $this->ary, 1));
    }

    public function testAdjacentDifference()
    {
        $this->assertEquals([], adjacentDifference([]));
        $this->assertEquals([4], adjacentDifference([4]));
        $this->assertEquals([2, 3, 2], adjacentDifference($this->ary));
    }

    public function testPartialSum()
    {
        $this->assertEquals([], partialSum([]));
        $this->assertEquals([2, 7, 14], partialSum($this->ary));
    }
}
?>
